<a href="https://colab.research.google.com/github/Jojomaque/Portfolio/blob/main/Jojo_Preparing_data_for_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Preparing Data for Analysis**



*   Exercise 1 - Checking for null vlues
*   Exercise 2 - Removing null values
*   Exercise 3 - Removing duplicate records from a data set
*   Exercise 4 - Normalising values where necessary
*   Exercise 5









### Exercise 1 - check for null values 
---

1. Read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv 
2. check if any NA values exist in the dataframe and print the result 
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071) 



In [ ]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv')
housing_london_original = df
housing_london_original.isnull().values.any()

True

In [ ]:
#.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071)
housing_london_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB


### Exercise 2 - remove null values 
---

1. Remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. Remove all NA values across whole dataframe 

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

In [ ]:
#Remove rows with NA values for life_satisfaction
housing_london_original.dropna(subset = ['life_satisfaction'])

code                  area  ... no_of_houses  borough_flag
613   E09000002  barking and dagenham  ...      71079.0             1
614   E09000003                barnet  ...     139346.0             1
615   E09000004                bexley  ...      95037.0             1
616   E09000005                 brent  ...     112083.0             1
617   E09000006               bromley  ...     135036.0             1
...         ...                   ...  ...          ...           ...
1010  E12000009            south west  ...          NaN             0
1013  E92000001               england  ...   24172166.0             0
1017  N92000002      northern ireland  ...          NaN             0
1018  S92000003              scotland  ...          NaN             0
1019  W92000004                 wales  ...          NaN             0

[352 rows x 12 columns]

In [ ]:
#Remove rows with NA values for whola dataframe
housing_london_original.dropna()

code                  area  ... no_of_houses  borough_flag
613   E09000002  barking and dagenham  ...      71079.0             1
614   E09000003                barnet  ...     139346.0             1
615   E09000004                bexley  ...      95037.0             1
616   E09000005                 brent  ...     112083.0             1
617   E09000006               bromley  ...     135036.0             1
...         ...                   ...  ...          ...           ...
999   E09000031        waltham forest  ...     103029.0             1
1000  E09000032            wandsworth  ...     146162.0             1
1001  E09000033           westminster  ...     124509.0             1
1008  E12000007                london  ...    3556161.0             0
1013  E92000001               england  ...   24172166.0             0

[267 rows x 12 columns]

## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])` 

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 3 - Removing duplicate entries 
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01* 

In [ ]:
#remove duplicate area entries keeping first instance
housing_london_original.drop_duplicates(subset = ['area'])


code                      area  ... no_of_houses  borough_flag
0   E09000001            city of london  ...          NaN             1
1   E09000002      barking and dagenham  ...          NaN             1
2   E09000003                    barnet  ...          NaN             1
3   E09000004                    bexley  ...          NaN             1
4   E09000005                     brent  ...          NaN             1
5   E09000006                   bromley  ...          NaN             1
6   E09000007                    camden  ...          NaN             1
7   E09000008                   croydon  ...          NaN             1
8   E09000009                    ealing  ...          NaN             1
9   E09000010                   enfield  ...          NaN             1
10  E09000011                 greenwich  ...          NaN             1
11  E09000012                   hackney  ...          NaN             1
12  E09000013    hammersmith and fulham  ...          NaN             1
13  E09000014                  haringey  ...          NaN             1
14  E09000015                    harrow  ...          NaN             1
15  E09000016                  havering  ...          NaN             1
16  E09000017                hillingdon  ...          NaN             1
17  E09000018                  hounslow  ...          NaN             1
18  E09000019                 islington  ...          NaN             1
19  E09000020    kensington and chelsea  ...          NaN             1
20  E09000021      kingston upon thames  ...          NaN             1
21  E09000022                   lambeth  ...          NaN             1
22  E09000023                  lewisham  ...          NaN             1
23  E09000024                    merton  ...          NaN             1
24  E09000025                    newham  ...          NaN             1
25  E09000026                 redbridge  ...          NaN             1
26  E09000027      richmond upon thames  ...          NaN             1
27  E09000028                 southwark  ...          NaN             1
28  E09000029                    sutton  ...          NaN             1
29  E09000030             tower hamlets  ...          NaN             1
30  E09000031            waltham forest  ...          NaN             1
31  E09000032                wandsworth  ...          NaN             1
32  E09000033               westminster  ...          NaN             1
33  E12000001                north east  ...          NaN             0
34  E12000002                north west  ...          NaN             0
35  E12000003  yorkshire and the humber  ...          NaN             0
36  E12000004             east midlands  ...          NaN             0
37  E12000005             west midlands  ...          NaN             0
38  E12000006                      east  ...          NaN             0
39  E12000007                    london  ...          NaN             0
40  E12000008                south east  ...          NaN             0
41  E12000009                south west  ...          NaN             0
42  E13000001              inner london  ...          NaN             0
43  E13000002              outer london  ...          NaN             0
44  E92000001                   england  ...          NaN             0
45  K02000001            united kingdom  ...          NaN             0
46  K03000001             great britain  ...          NaN             0
47  K04000001         england and wales  ...          NaN             0
48  N92000002          northern ireland  ...          NaN             0
49  S92000003                  scotland  ...          NaN             0
50  W92000004                     wales  ...          NaN             0

[51 rows x 12 columns]